Добавлена работа с базой данных H2

YetiSPb opened the pull request:
Добавил базу H2
Подключил добавил новые эндпоинты

Took 11 hours 37 minutes

YetiSPb committed:
Добавлена работа с базой данных H2

Добавил базу H2
Подключил добавил новые эндпоинты

Took 11 hours 37 minutes

YetiSPb committed:
Добавлена работа с базой данных H2

Добавил базу H2
Подключил добавил новые эндпоинты

Took 4 minutes

YetiSPb committed:
Добавлена работа с базой данных H2

Добавил базу H2
Подключил добавил новые эндпоинты

Took 3 minutes

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/log/Logger.java:
```diff
@@ -0,0 +1,16 @@
+@Slf4j
+public class Logger {
+
+    public static void logRequest(HttpMethod method, String uri, String body) {
```
Интересный вариант работы с логом!

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,61 +1,91 @@
+
+    public Collection<Film> getFilms() {
+        Collection<Film> filmsInStorage = filmStorage.getFilms();
+        Logger.logSave(HttpMethod.GET, "/films", filmsInStorage.toString());
```
Не очень правильно привязывать методы сервиса к http-методам - к использованию http привязаны только контроллеры, а сервисы содержат бизнес-логику и могут работать и без использования http, через какой-то другой интерфейс (относится ко всему файлу и соседним файлам)

YetiSPb replied:
Не понял. Поясни пожалуйста поточнее. Что к чему привязывать на курсе не давали.

annaalkh replied:
Общая структура должна быть такая: storage/dao - хранение данных, сервисы - бизнес-логика, контроллеры - взаимодействие с пользователем (например, через rest). Соответственно, как именно с пользователем взаимодействуем - должно быть полностью скрыто в слое контроллеров, а сервисы должны отвечать только за сам алгоритм работы, условно - та же инкапсуляция, но на более высоком уровне

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,61 +1,91 @@
 
-        if (count > popularFilms.size()) {
-            count = popularFilms.size();
+    private Film checkValidation(Film film) {
```
Обычно подобные валидационные методы не возвращают значения (возвращают void)

YetiSPb replied:
Можно конечно, но это увеличит количество строк.

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,61 +1,91 @@
-        filmStorage.getFilmById(id).addLike(userId);
+    public void addLike(long id, long userId) {
+        boolean addition;
+        filmStorage.getFilmById(id);
```
Такой вызов метода выглядит не очень хорошо и вводит в заблуждение - метод getFilmById рассчитан на то, чтобы получить фильм и как-то его дальше использовать, а здесь полученное значение просто игнорируется. Если цель вызова - проверить существование пользователя, правильнее будет создать отдельный метод в filmStorage, например checkFilmExists

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,61 +1,91 @@
+    public void addLike(long id, long userId) {
+        boolean addition;
+        filmStorage.getFilmById(id);
+        userService.getUserById(userId);
```
Здесь тоже - стоит использовать другой метод

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,61 +1,91 @@
-        filmStorage.getFilmById(id).deleteLike(userId);
+    public void unlike(long id, long userId) {
+        boolean removal;
+        filmStorage.getFilmById(id);
```
И здесь

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,61 +1,91 @@
-        Collections.reverse(popularFilms);
+    public List<Film> getTheBestFilms(int count) {
+        List<Film> bestFilms = likesStorage.getTheBestFilms(count).stream()
+                .map(filmStorage::getFilmById)
```
Не самый эффективный метод реализации - на каждый фильм будет выполняться дополнительный запрос, а каждый запрос - это нагрузка на бд. Правильнее было бы выбрать все необходимые данные сразу одним запросом (используя для этого join)

YetiSPb replied:
Я понимаю, но времени нет. Так что что есть  о есть.

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -0,0 +1,40 @@
+
+    public List<Genre> getListOfGenres(long id) {
+        List<Genre> genreList = filmGenreLineStorage.getListOfGenres(id).stream()
+                .map(genreStorage::getGenreById)
```
Здесь тоже - лучше выбрать все данные сразу, а не сначала id, потом все остальное

YetiSPb replied:
Я понимаю, но всё так же из за отставания после переписывания первого проекта я месяц SQL вложил в две недели.

annaalkh replied:
Ок - на данный момент это не критично, но в следующих работах станет более важным

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,65 +1,95 @@
-        List<User> userFriends = getAllUserFriends(idUser);
-        List<User> friendFriends = getAllUserFriends(idFriends);
+    public List<User> getListOfFriends(long id) {
+        userStorage.getUserById(id);
```
Тоже вызов метода без использования результата

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,65 +1,95 @@
+    public List<User> getListOfFriends(long id) {
+        userStorage.getUserById(id);
+        List<User> friendList = friendsStorage.getListOfFriends(id).stream()
+                .map(userStorage::getUserById)
```
Нужно использовать один запрос

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,65 +1,95 @@
-        List<User> commonFriends = new ArrayList<>();
+    public List<User> getAListOfMutualFriends(long id, long otherId) {
+        userStorage.getUserById(id);
+        userStorage.getUserById(otherId);
```
И здесь

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,65 +1,95 @@
+        userStorage.getUserById(id);
+        userStorage.getUserById(otherId);
+        List<User> mutualFriends = friendsStorage.getAListOfMutualFriends(id, otherId).stream()
+                .map(userStorage::getUserById)
```
И здесь - один запрос

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,65 +1,95 @@
-            if (friendFriends.contains(user)) {
-                commonFriends.add(user);
-            }
+    private User checkValidation(User user) {
```
Убрать возвращаемое значение

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -0,0 +1,119 @@
+public class FilmDbStorage implements FilmStorage {
+    private final JdbcTemplate jdbcTemplate;
+    private final LikesStorage likesStorage;
+    private final MpaService mpaService;
```
Не очень правильно, когда в storage используется сервис - сервисы это более высокий уровень приложения, из storage стоит использовать только другие storage

YetiSPb replied:
Если бы про это рассказывали до того как давали ТЗ то и реализация была бы чётче.

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -0,0 +1,119 @@
+                .releaseDate(resultSet.getDate("release_date").toLocalDate())
+                .duration(resultSet.getInt("duration"))
+                .rate(resultSet.getInt("rate"))
+                .mpa(mpaService.getMpaById(resultSet.getInt("mpa_id")))
```
Здесь та же проблема - на каждый фильм будет выполняться несколько дополнительных запросов, такого лучше избегать

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -0,0 +1,119 @@
+
+
+        Collection<Genre> existingGenres = oldFilm.getGenres();
+        if (existingGenres != null && !existingGenres.isEmpty()) {
```
Было бы лучше, если бы existingGenres никогда не содержала null - в коде всегда стоит избегать случаев, когда коллекция равна null, например можно инициализировать ее сразу при объявлении поля в Film

YetiSPb replied:
Я страху перед null в Java не привык. У меня не было таких проблем. Взяли, проверили и поехали дальше.

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmGenreLineDbStorage.java:
```diff
@@ -0,0 +1,60 @@
+        List<Genre> uniqueGenres = genres.stream()
+                .distinct()
+                .collect(Collectors.toList());
+        getJdbcTemplate().batchUpdate(sqlQuery, new BatchPreparedStatementSetter() {
```
Использовать batchUpdate - очень правильно!

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/LikesDbStorage.java:
```diff
@@ -0,0 +1,62 @@
+                "left outer join LIKES ON FILMS.FILM_ID = LIKES.FILM_ID " +
+                "group by FILMS.FILM_id " +
+                "order by count(distinct LIKES.USER_ID) desc " +
+                "limit + ?";
```
А + зачем?

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -0,0 +1,119 @@
+        SimpleJdbcInsert simpleJdbcInsert = new SimpleJdbcInsert(jdbcTemplate)
+                .withTableName("films")
+                .usingGeneratedKeyColumns("film_id");
+        long filmId = simpleJdbcInsert.executeAndReturnKey(toMap(film)).longValue();
```
Отлично, что ты использовал executeAndReturnKey

annaalkh commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,22 @@
+DELETE FROM LIKES;
```
Удалять данные здесь нельзя - файлы data.sql и schema.sql выполняются каждый раз при старте приложения, если удалять в них какие-либо данные, и использовании бд вообще теряется смысл

YetiSPb replied:
Полностью согласен. Только создатель тестов это не учёл.

annaalkh replied:
Учел - тесты прекрасно проходят без подобных конструкций, уже проверено с другими студентами. Это нужно исправить

annaalkh commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,22 @@
+DELETE FROM FRIENDS;
+DELETE FROM USERS;
+DELETE FROM FILMS;
+ALTER TABLE USERS ALTER COLUMN USER_ID RESTART WITH 1;
```
К этим строчкам относится то же самое

YetiSPb replied:
Тут та же проблема с тестами что и выше.

annaalkh commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,60 @@
+CREATE TABLE IF NOT EXISTS users
+(
+    user_id  Integer GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
```
Имена типов обычно пишут в uppercase

annaalkh commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,60 @@
+CREATE TABLE IF NOT EXISTS users
+(
+    user_id  Integer GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
+    email    varchar,
```
При создании varchar-полей всегда рекомендуется указывать их размер - это влияет на эффективное использование памяти бд и производительность ее работы

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/model/FilmGenreLine.java:
```diff
@@ -0,0 +1,11 @@
+
+@Data
+@Builder
+public class FilmGenreLine {
```
Обычно для таких вспомогательных связей отдельные сущности не создают - можно обойтись существующими (также как можно обойтись и без отдельного storage для них)

YetiSPb replied:
Спасибо. Я приму это к сведению.

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/model/Likes.java:
```diff
@@ -0,0 +1,11 @@
+
+@Data
+@Builder
+public class Likes {
```
И этот

YetiSPb replied:
Спасибо. Я приму это к сведению.

YetiSPb replied:
Используется в LikesDbStorage

annaalkh commented on src/main/java/ru/yandex/practicum/filmorate/model/Friends.java:
```diff
@@ -0,0 +1,11 @@
+
+@Data
+@Builder
+public class Friends {
```
Этот класс, как я вижу, нигде не используется - он тоже выглядит лишним

YetiSPb replied:
Используется в FriendsDbStorage

annaalkh reviewed (commented):
Евгений, привет! В целом - отлично! Некоторые моменты очень понравились. Есть и отдельные замечания - особенно рекомендую обратить внимание на тот момент, что стоит избегать дополнительных запросов к бд там, где это возможно (подробнее про эту проблему можно почитать, например, здесь https://habr.com/ru/companies/otus/articles/529692/)

YetiSPb committed:
Добавлена работа с базой данных H2

Поправил коментарии
Но не исправлял работу с запросами. Времени катострофически не хватает. Я понимаю что плучадть данные лучше одним запросом, а не десятью. Спокойно оптимихарую на каникулах.

Took 2 hours 52 minutes

YetiSPb committed:
Добавлена работа с базой данных H2

Поправил коментарии
Но не исправлял работу с запросами. Времени катострофически не хватает. Я понимаю что плучадть данные лучше одним запросом, а не десятью. Спокойно оптимихарую на каникулах.

Took 4 minutes

YetiSPb committed:
Добавлена работа с базой данных H2

Поправил коментарии
Но не исправлял работу с запросами. Времени катострофически не хватает. Я понимаю что плучадть данные лучше одним запросом, а не десятью. Спокойно оптимихарую на каникулах.

Took 33 minutes

annaalkh reviewed (commented):
Удаление данных в sql-файлах нужно исправить

YetiSPb committed:
Добавлена работа с базой данных H2

Took 26 minutes

annaalkh reviewed (approved):
Теперь хорошо!
